validator_client/src/beacon_node_fallback.rs

```diff
@@ -141,114 +129,167 @@ impl<T> Errors<T> {
 }
 
 /// Reasons why a candidate might not be ready.
-#[derive(Debug, Clone, Copy)]
+#[derive(Debug, Clone, Copy, PartialEq, Deserialize, Serialize)]
 pub enum CandidateError {
+    PreGenesis,
     Uninitialized,
     Offline,
     Incompatible,
-    NotSynced,
+    TimeDiscrepancy,
+}
+
+impl std::fmt::Display for CandidateError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            CandidateError::PreGenesis => write!(f, "PreGenesis"),
+            CandidateError::Uninitialized => write!(f, "Uninitialized"),
+            CandidateError::Offline => write!(f, "Offline"),
+            CandidateError::Incompatible => write!(f, "Incompatible"),
+            CandidateError::TimeDiscrepancy => write!(f, "TimeDiscrepancy"),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct CandidateInfo {
+    pub index: usize,
+    pub endpoint: String,
+    pub health: Result<BeaconNodeHealth, CandidateError>,
+}
+
+impl Serialize for CandidateInfo {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("CandidateInfo", 2)?;
+
+        state.serialize_field("index", &self.index)?;
+        state.serialize_field("endpoint", &self.endpoint)?;
+
+        // Serialize either the health or the error field based on the Result
+        match &self.health {
+            Ok(health) => {
+                state.serialize_field("health", health)?;
+            }
+            Err(e) => {
+                state.serialize_field("error", &e.to_string())?;
+            }
+        }
+
+        state.end()
+    }
 }
 
 /// Represents a `BeaconNodeHttpClient` inside a `BeaconNodeFallback` that may or may not be used
 /// for a query.
+#[derive(Clone, Debug)]
 pub struct CandidateBeaconNode<E> {
-    beacon_node: BeaconNodeHttpClient,
-    status: RwLock<Result<(), CandidateError>>,
+    pub index: usize,
+    pub beacon_node: BeaconNodeHttpClient,
+    pub health: Arc<RwLock<Result<BeaconNodeHealth, CandidateError>>>,
```

Comment:
Is this `RwLock` necessary given that `candidates` is already behind a `RwLock` in `BeaconNodeFallback`?

https://github.com/sigp/lighthouse/blob/7eab573909976b6ee5798840f04e9639ce764854/validator_client/src/beacon_node_fallback.rs#L375

Reply:
ah never mind, I see that it's intentional from its usage in `update_all_candidates`.